### ISSUE TITLE
ENHANCEMENT: Add instrument resolutions of specific instruments

### ISSUE THREAD

**Issue by RastislavTuranyi** [ISISNeutronMuon/MDANSE#76]
Another thing that was brought to me during NMSUM. Currently, the only way to input an instrument resolution to MDANSE analyses is by manually typing in the parameters of the Gaussian. This is often not trivial, especially for external users who do not have this information at hand, and so might have to contact an instrument scientist or do extensive searches. Overall, this wastes everyone's time and makes for a bad user experience.

Therefore, a drop-down box should be implemented that allows users to select a specific instrument, the resolution of which would be automatically filled in, similar to how it is in Mantid.

**Comment by gonzalezma** [comment]
Are we sure about this? I will do the old grumpy guy here, but I think that we provide a tool and we can expect some effort from the users. My personal preference is to keep things as simple as possible, and I don't see any reasonable and simple way of handling this. My arguments:

1. There are many instruments that do not have a single resolution function (e.g. in IN5 the number of possibilities is almost infinite, and surely there are similar instrument at ISIS). And the number of instruments in the world is huge.  Are we sure that we want to add a huge list of instruments with their possible variants and keep adding new instruments on demand?

2.  The resolution is handled in time (i.e. we compute the FT of I(t)*R(t)), so any function proposed in frequency should have a well defined FT in time. Can we ensure this?

My personal experience is that in most cases the resolution can be well approximated by a Gaussian (and only rarely one may have a Lorentzian or pseudo-Voigt), but I understand that this is not always the case, and some instruments have non-symmetric resolutions. Would it be relatively simple to allow using a sum of Gaussians to try to handle in an approximate way those cases?

As for the external users, anyone who has done a neutron experiment needs the resolution function to analyse his/her data, so I don't agree that they don't have this information at hand.

I know that I'm being quite conservative here, so please give your contrary opinions and suggestions.

**Comment by eurydice76** [comment]
I find the argument of Miguel very convincing.

**Comment by RastislavTuranyi** [comment]
I also have to say that I am convinced. Personally, I have never performed a neutron experiment and have not used MDANSE as a user, so I can only relay others' experiences. Therefore, I'd say that we can close this issue, but I can try to contact the user who brought this up with me if you would like to know more than I presented in my initial post. Also, @sanghamitra-mukhopadhyay, what do you think about this?

**Comment by ajjackson** [comment]
Some existing codes implement resolution functions: AbINS ships resolution functions for 1-D instruments (TOSCA, ILL-LAGRANGE) and uses PyChop to calculate energy resolution for ISIS direct-geometry instruments MAPS, MARI and MERLIN in the soon-to-be-released Abins2D mode. (There is also a fitted function for ILL-PANTHER, and I have recently heard from a user about more sophisticated options there.) I have developed an approximate method method for applying energy-dependent broadening functions at low computational cost; an initial version of this is already in AbINS and an improved version (developed as an SCD graduate project) is in Euphonic for adaptive phonon DOS calculations.

Horace implements difficult and expensive 4-D resolution functions by Monte Carlo sampling, and we have a couple of parallel projects at different stages looking into alternative ways of doing that. Meanwhile Horace is going to lean on the new Euphonic library for some phonon purposes: this library also provides some plotting tools with simple broadening options, which are likely to be developed to accept more complex functions. However, the general consensus in the Euphonic team is that we don't want to fill that library with instrument parameters - it should be a reasonably pure tool for force-constants manipulation and scattering function calculations.

I wonder if a natural way forwards is to develop a separate Python library that implements instrument resolution functions. Then there is some kind of authoritative open-source reference that can be used by many codes, including new ones. I suppose it would be a kind of successor to PyChop?

It is very likely that such a dependency _would_ rely on modern Python versions, Scipy etc. which could present a barrier to use in the current MDANSE. But it seems like the more sustainable way in the long run. (And conveniently provides an excuse to defer the work...)

**Comment by mducle** [comment]
I think that a resolution library with common instruments included would be a good idea. PyChop in Mantid only handles the time-resolution but there is also in principle spatial broadening due to the finite size of the beam, sample and detectors which has correlations with and affects the final measured time/energy widths even if you've got dispersionless modes. For coherent excitations which AbINS2D would handle which are dispersive the spatial resolution is then also important.

For reactor sources at all energies and pulse sources at high energies the Gaussian approximation is good enough and in this case the resolution function should be analytic so I think a FT from frequency to time is well defined. However, for pulsed sources the time structure of the moderator pulse becomes progressively more assymmetric with longer times which gives distinctly non-Gaussian energy line-shapes at medium-to-low energies. For direct geometry instruments you can spin the choppers faster (up to a limit!) to cut out the assymmetric tails at the cost of incident flux but some people will chose not to do this (or have too small samples to allow this). I guess in this case we can calculate the time lineshape directly for MDANSE. For other AbINS/euphonic or other programs which compute in (Q,E) space we already have a program (Horace as Adam mentioned) to compute the lineshape in energy.

In any case these are computations are specialised enough that I think a separate library would make more sense than to incorporate it into MDANSE or AbINS etc directly (it might make sense in Mantid, but Mantid is already too big in my view, and is moving towards modularisation in any case; there was actually a [C++ implementation](https://docs.mantidproject.org/v4.0.0/algorithms/SimulateResolutionConvolvedModel-v1.html) in Mantid but it was unused, not well tested and removed a few years ago...)

**Comment by sanghamitra-mukhopadhyay** [comment]
Yes, at this moment MDANSE is using just Gaussian broadening as a simplest way and agree @gonzalezma  and @eurydice76 's point of keeping it simple. 
However, as we do this  in abINS as mentioned by @ajjackson , it is a long term users requirements and one of the objective of the MDANSE project is to include the instrument specific resolutions to compare directly with experiments. So in this context, @RastislavTuranyi I think we should keep this issue alive and try to implement this in near future. 
It is a good idea to keep it as a separate library as @mducle mentioned, so that relevant codes can use that library as they prefer. At present since there is no such library, if we can separate out the code used in pyChop at Mantid and integrate it in MDANSE, or just  take the output of PyChop as a data file and put it in MDANSE, that will be great. In the later case, we just have to keep the option in MDANSE to use users-supplied resolution function as a ascii data file with specific data format. 
For a library we have to ensure how do we implement the code to facilitate requirements for abINS, PyChop and MDANSE, because those process data in different ways. In case of MDANSE it would be relevant to apply the information in the time domain, because it deals with trajectories, On the other hand abINS takes input in the form of frequency. 
To keep life simple as mentioned above, we can take this resolution information in the Q-E domain as obtained from  PyChop, which will be used by MDANSE  the same way as in abINS.

**Comment by ajjackson** [comment]
The library as I envision it would be able to take data as an array of S(ω), S(|q|,ω), S(**q**,ω) [or some other property along those axes] and return a broadened array of the same or lower (if collapsing from 4-D) shape.

The functions to compute scalar std/FWHM or covariance matrix as a function of (*q*,ω) would also be exposed through the API in case people want to implement the broadening some other way (e.g. to combine it with other contributions and calculate in a single shot).

**Comment by mducle** [comment]
So actually the way that PyChop and Horace compute the instrument resolution is actually as distributions in the time and spatial coordinates of the instrument components. PyChop makes the Gaussian approximation and computes a width sigma for each component's contribution and these are added in quadrature and the resulting time width is converted into energy. In Horace, points are randomly sampled from the time-spatial distributions and converted into (**q**,ω) for a given (single-crystal) sample orientation and lattice parameters; a model S(**q**,ω) is evaluated at N of these points for a given detector element and averaged.

So, for MDANSE I think we can output the time (and **q**?) distributions R(t) at the sample position which can be directly used in the resolution convolution calculation.

**Comment by sanghamitra-mukhopadhyay** [comment]
Yes we do measure resolutions at the beginning of each cycle with vanadium and sometime for a particular experiment on QENS instruments, IRIS and OSIRIS. This is done for other direct energy spectroscopic instruments, MERLIN and MAPS. However, that is not the norm for TOSCA, where knowledge from old measurements done at the time of instrument upgradations has been taken as resolution.

**Comment by ajjackson** [comment]
How much does it change, if at all?

**Comment by AnthonyLim23** [comment]
As a "quick" way to fix it the user could provide the values (as they do now) or a file with the values. Then we just need to provide the user with the correct file. 

A library is probably a better fix long term but will take time.

**Comment by mducle** [comment]
> How much does it change, if at all?

For the calibrations at the start of the cycle which are done with standard settings and for vanadiums on the QENS instruments I don't think it changes much but it's still done in order to calibrate and check the instrument*.

For the direct geometry instruments because you can set the choppers to give you different neutron energies you should in principle measure a vanadium for each setting you used. This not only gives you an estimate of the elastic energy width** but also a normalisation factor to turn counts into cross-sections [we usually use milibarns/steradian/meV].

Also, the vanadium width measurement is a good check on the resolution calculations because the calculated widths tend to _underestimate_ the measured widths (by ~10-20%), at least with the direct geometry instruments... I'm not sure why this is. It could be due to physical imperfections in the choppers (e.g. for the Fermi choppers [some of] the channels could be slightly wider than in the drawings; for disk choppers this is not sure an issue I think); or our model of the moderator pulse time structure is not good (it's based on MCMP Monte Carlo particle transport calculations); or due to underestimating the effect of the supermirror guides that are in most instruments (we don't really model these properly with raytracing at the moment but just estimate their effect empirically).

----

*Often it's used to check that all the detectors are still ok (they can fail or glitch etc between cycles or even during a cycle). Also the accelerator could be running at a different proton energy so giving you more or fewer neutrons; so if you want to compare / combine data from different cycles you'd need this to renormalise... if you compare the measured count rate with previous it would also give you an indication of whether there's an issue with the guides or other components upstream of the sample if you see a significant drop in count rate for example (this was the first indication on MAPS that there was some obstruction in the shutter which led to it being replace...; also it's known that supermirror guides can sometimes delaminate so you'd want to know if that happens...)

**The energy width changes with energy transfer for direct geometry, not so much for indirect geometry instruments so you can approximate it as constant on e.g. TOSCA, IRIS and OSIRIS.

**Comment by ajjackson** [comment]
> As a "quick" way to fix it the user could provide the values (as they do now) or a file with the values. Then we just need to provide the user with the correct file.

In Euphonic we offer kinematic constraints as an option. Although they have to be provided as parameters, we provide a table in the docs with some suggested values for known instruments.
https://euphonic.readthedocs.io/en/latest/spectra.html#kinematic-constraints

**Comment by gonzalezma** [comment]
As usual, many visible and hidden problems lie on the interface. My personal preference is to keep problems separate and accept that MDANSE mission is to work on the simulation side. From this point of view, I think that providing a general set of standard mathematical functions that approximate the instrument resolution is enough in most cases. Most of the time, when comparing experiment and simulation, the main differences come from the model (the difference between the real sample measured and the computer model) and the approximation in the simulation (validity of the potential and/or DFT limitations), so having a perfect match of the instrument resolution is not essential.

For the cases where a more refined resolution function is needed, this could be done directly by the user or using the suggested library or some of the "experimental" software mentioned in the discussion (Mantid, Euphonics, Horace, etc.). At present MDANSE does not read experimental data (e.g NeXus) nor user files, which would be needed to have a complete instrument description or use a measured data file as resolution. This could be added of course, but I'm not sure that this is the best approach.

**Comment by ajjackson** [comment]
If it is easy for users to get 2-D data file outputs from MDANSE, without any broadening and in a reasonable format, then said resolution library could include a command-line tool for visualising/writing the data with different resolution functions.

4-D data is a bit more tricky but the same principles should apply.

It's good to decouple expensive calculations from messing around with resolution, anyway. The only problem is that it could be more difficult to educate users to work this way.
